fms/modules/feedforward.py

```diff
@@ -658,16 +743,32 @@ def reset_parameters(self):
 
     def forward(self, x: torch.Tensor) -> torch.Tensor:
         B, S = x.shape[:2]
+
         x = x.view(-1, self.dim)
-        # T = num_tokens, E = num_experts, D = hidden dim, A = activated experts
-        # x: [T, D]
-        scores = self.gate(x)  # [T, E]
-        expert_weights = F.softmax(scores, dim=-1)
-        expert_weights, expert_indices = torch.topk(
-            expert_weights, self.num_activated_experts, dim=-1
-        )  # [T, A], [T, A]
-        expert_weights /= expert_weights.sum(dim=-1, keepdim=True)  # [T, A]
-        expert_outs = self.cond_ffn(x, expert_indices)
-        int_v1 = torch.einsum("tai,ta -> ti", expert_outs, expert_weights)
-        int_v2 = int_v1.view(B, S, self.dim)
+        if self.use_bias:
+            scores = x.to(self.gate.weight.dtype)
+            g = self.gate(scores)
+            experts = torch.topk(g, k=self.num_activated_experts, dim=-1, sorted=True)
+            expert_weights = F.softmax(experts.values, dim=1)
+            expert_indices = experts.indices
+            expert_outs = self.cond_ffn(x, expert_indices, expert_weights, scores)
+
+        else:
+            # T = num_tokens, E = num_experts, D = hidden dim, A = activated experts
+            # x: [T, D]
+            scores = self.gate(x)  # [T, E]
+            expert_weights = F.softmax(scores, dim=-1)
+            expert_weights, expert_indices = torch.topk(
+                expert_weights, self.num_activated_experts, dim=-1
+            )  # [T, A], [T, A]
+            expert_weights /= expert_weights.sum(dim=-1, keepdim=True)  # [T, A]
+
+            expert_outs = self.cond_ffn(x, expert_indices, expert_weights, scores)
+            int_v1 = torch.einsum("tai,ta -> ti", expert_outs, expert_weights)
+
+        if self.use_bias:
+            int_v2 = expert_outs.view(B, S, self.intermediate_size)
+        else:
+            int_v2 = int_v1.view(B, S, self.dim)
```

Comment:
Understanding check: is the only intended difference between these two code paths that one path uses one or more bias tensors and the other does not? If so, I am confused why the two paths look so different.

I also don't see anything in either block which references a bias tensor, which is also confusing to me.

If we do need two code paths here, we could also merge these two separate if-blocks into a single one.